Copy edits (#273)
CodeYourFuture/curriculum: content/js2/blocks/multiple-params/index.md

We've already worked out how to update the query params object given a **single key-value pair** in the query string.

To work out our strategy, let's consider what we already know how to do. We already know how to take a key-value pair as a string, and add it to our object.

> 💡 Key insight: If we can do it for **one pair**, we can try doing it for a **list of pairs** too.

> So we're missing a step - breaking up the string of lot sof key-value pairs into an array where each element is a single key-value pair. If we do this, then we can iterate over the array, and do what we already know how to do on each key-value pair.

Reviewer: ```suggestion
So we're missing a step - breaking up the string of lots of key-value pairs into an array where each element is a single key-value pair. If we do this, then we can iterate over the array, and do what we already know how to do on each key-value pair.
```

function parseQueryString(queryString) {
  // suppose queryString has a value of "sort=lowest&colour=yellow"
  const queryParams = {};
  const keyValuePairs = queryString.split("&"); // keyValuePairs will point to ["sort=yellow", "colour=yellow"]
}
```

#### 🎯 Sub-goal 2: add each key-value pair in the array to the query params object

> Once we've got an array we can iterate through the key-value pairs and update the `queryParams` object each time (just like we did when we just had one key-value pair).

Reviewer: ```suggestion
Once we've got an array we can iterate through the key-value pairs and update the `queryParams` object each time (like we did when we just had one key-value pair).
```

{{</tab>}}

{{</tabs>}}

Now that we've worked out how to solve this problem in the case of multiple query parameters, let's integrate that solution into our previous implementation, to make sure it works for all cases.

> We can keep our `if (queryString.length === 0) {` check from before. We don't need to do anything special for the one-value case, as an array containing one element gets iterated the same as an array multiple elements:

Reviewer: ```suggestion
We can keep our `if (queryString.length === 0) {` check from before. We don't need to do anything special for the one-value case, as an array containing one element gets iterated the same as an array of multiple elements:
```

```js
function parseQueryString(queryString) {
  const queryParams = {};
  if (queryString.length === 0) {
    return queryParams;
  }
  const keyValuePairs = queryString.split("&");

  for (const pair of keyValuePairs) {
    const [key, value] = pair.split("=");
    queryParams[key] = value;
  }

  return queryParams;
}
```

When we're solving problems involving several values, often we need slightly differently handling for the cases when there are 0, 1, or more than 1 values. In our example here, we need to treat 0 values specially (if the query string is empty, we return early), but we can handle 1 and more than 1 the same way.

> When you're breaking down problems, think to yourself: What are special cases we may need to handle differently?

Author: @Dedekind561

Can we also talk about edge-cases here? Three important ones to identify are missing =s (`foo&bar=baz`), repeated keys (`foo=bar&foo=baz`), and empty segments (`?&foo=bar`). I think we should probably have a section on identifying the edge-cases, then set an exercise to work out how to handle them, and implement that?

Reviewer: I'm afraid of there being too much content in the prep. Can we move all the edge case material into a separate exercise in the coursework? Or even do a workshop to explore edge cases and how to handle them?

Author: I'm happy to skip the exercise, but I'd like to keep the prompt to think about them... What do you think?

Reviewer: Yes absolutely - I like the prompt
